File: tools/auction-arbitrage/arb/valuation/ebay.py
"""eBay sold/completed listings as comps.

eBay's sold-listing search page is server-rendered HTML that we can parse without an API key.
This is best-effort: eBay may rate-limit or serve a challenge page; on any failure we return [].
If you have eBay developer keys you can swap in the Browse/Marketplace Insights APIs here.
"""
from __future__ import annotations

import html
import logging
import re
import statistics
import time
import urllib.parse
from typing import Any

import httpx

from .base import Comp

log = logging.getLogger(__name__)

UA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/128.0 Safari/537.36"
_ITEM = re.compile(r'<li class="s-item[^"]*"[^>]*>(.*?)</li>', re.S)
_TITLE = re.compile(r'<(?:div|span|h3) class="s-item__title[^"]*"[^>]*>(.*?)</(?:div|span|h3)>', re.S)
_PRICE = re.compile(r'<span class="s-item__price"[^>]*>(.*?)</span>', re.S)
_LINK = re.compile(r'<a class="s-item__link"[^>]*href="([^"]+)"')
_DATE = re.compile(r'(?:Sold|Ended)\s+([A-Z][a-z]{2}\s+\d{1,2},\s+\d{4})')
_TAG = re.compile(r"<[^>]+>")
_MONEY = re.compile(r"[\d,]+(?:\.\d{2})?")


def sold_search_url(query: str) -> str:
    q = urllib.parse.quote_plus(query)
    return f"https://www.ebay.com/sch/i.html?_nkw={q}&LH_Sold=1&LH_Complete=1&_sop=13&_ipg=60"


def clean_query(title: str, max_words: int = 8) -> str:
    t = re.sub(r"\(.*?\)|\[.*?\]", " ", title or "")
    t = re.sub(r"[^A-Za-z0-9\-\.\s]", " ", t)
    words = [w for w in t.split() if len(w) > 1]
    junk = {"lot", "of", "the", "and", "with", "new", "nib", "nwt", "untested", "as", "is", "read"}
    words = [w for w in words if w.lower() not in junk]
    return " ".join(words[:max_words])


def fetch_sold_comps(query: str, *, limit: int = 30, timeout: float = 20.0) -> list[Comp]:
    if not query.strip():
        return []
    url = sold_search_url(query)
    try:
        r = httpx.get(url, headers={"user-agent": UA, "accept-language": "en-US,en;q=0.9"}, timeout=timeout,
                      follow_redirects=True)
        if r.status_code != 200 or "s-item" not in r.text:
            log.info("ebay sold search: no parsable results (%s)", r.status_code)
            return []
        body = r.text
    except httpx.HTTPError as e:
        log.info("ebay sold search failed: %s", e)
        return []
    comps: list[Comp] = []
    for block in _ITEM.findall(body):
        t = _TITLE.search(block)
        p = _PRICE.search(block)
        if not t or not p:
            continue
        title = html.unescape(_TAG.sub("", t.group(1))).strip()
        if title.lower().startswith("shop on ebay"):
            continue
        price_txt = html.unescape(_TAG.sub("", p.group(1)))
        m = _MONEY.search(price_txt)
        if not m:
            continue
        price = float(m.group(0).replace(",", ""))
        link = _LINK.search(block)
        d = _DATE.search(html.unescape(_TAG.sub(" ", block)))
        comps.append(Comp(title=title, price=price, source="ebay_sold",
                          url=(link.group(1).split("?")[0] if link else url), date=d.group(1) if d else ""))
        if len(comps) >= limit:
            break
    return comps


def summarize(comps: list[Comp]) -> dict[str, Any]:
    prices = sorted(c.price for c in comps if c.price > 0)
    if not prices:
        return {"n": 0}
    q = statistics.quantiles(prices, n=4) if len(prices) >= 4 else [prices[0], statistics.median(prices), prices[-1]]
    return {"n": len(prices), "p25": q[0], "median": statistics.median(prices), "p75": q[-1],
            "min": prices[0], "max": prices[-1]}


def polite_sleep(seconds: float = 1.5) -> None:
    time.sleep(seconds)

File: lib/spread/valuation/ebay.ts
// eBay comps. Two paths:
//  1. Sold/completed listings page (no key; server-rendered HTML). eBay blocks datacenter IPs often,
//     so on Vercel this is best-effort and returns [] on any trouble.
//  2. Browse API (official, free key: EBAY_CLIENT_ID + EBAY_CLIENT_SECRET). Returns ACTIVE listings
//     (asking prices, not sold) — weaker evidence, labelled as such, but reliable from a server.
import type { Comp } from "../types";
import { stripPrefix } from "./base";

const UA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/128.0 Safari/537.36";
const ITEM = /<li class="s-item[^"]*"[^>]*>([\s\S]*?)<\/li>/g;
const TITLE = /<(?:div|span|h3) class="s-item__title[^"]*"[^>]*>([\s\S]*?)<\/(?:div|span|h3)>/;
const PRICE = /<span class="s-item__price"[^>]*>([\s\S]*?)<\/span>/;
const LINK = /<a class="s-item__link"[^>]*href="([^"]+)"/;
const DATE = /(?:Sold|Ended)\s+([A-Z][a-z]{2}\s+\d{1,2},\s+\d{4})/;
const TAG = /<[^>]+>/g;
const MONEY = /[\d,]+(?:\.\d{2})?/;

function unescape(s: string): string {
  return s.replace(/&amp;/g, "&").replace(/&lt;/g, "<").replace(/&gt;/g, ">").replace(/&quot;/g, '"').replace(/&#39;/g, "'").replace(/&nbsp;/g, " ");
}

export function soldSearchUrl(query: string): string {
  return `https://www.ebay.com/sch/i.html?_nkw=${encodeURIComponent(query)}&LH_Sold=1&LH_Complete=1&_sop=13&_ipg=60`;
}

export function cleanQuery(title: string, maxWords = 8): string {
  const junk = new Set(["lot", "of", "the", "and", "with", "new", "nib", "nwt", "untested", "as", "is", "read"]);
  const t = stripPrefix(title).replace(/\(.*?\)|\[.*?\]/g, " ").replace(/[^A-Za-z0-9\-.\s]/g, " ");
  return t.split(/\s+/).filter((w) => w.length > 1 && !junk.has(w.toLowerCase())).slice(0, maxWords).join(" ");
}

export async function fetchSoldComps(query: string, limit = 30): Promise<Comp[]> {
  if (!query.trim()) return [];
  const url = soldSearchUrl(query);
  let body: string;
  try {
    const res = await fetch(url, { headers: { "user-agent": UA, "accept-language": "en-US,en;q=0.9" }, signal: AbortSignal.timeout(15000), cache: "no-store" });
    if (!res.ok) return [];
    body = await res.text();
    if (!body.includes("s-item")) return [];
  } catch {
    return [];
  }
  const comps: Comp[] = [];
  for (const m of body.matchAll(ITEM)) {
    const block = m[1];
    const t = TITLE.exec(block), p = PRICE.exec(block);
    if (!t || !p) continue;
    const title = unescape(t[1].replace(TAG, "")).trim();
    if (title.toLowerCase().startsWith("shop on ebay")) continue;
    const money = MONEY.exec(unescape(p[1].replace(TAG, "")));
    if (!money) continue;
    const link = LINK.exec(block);
    const d = DATE.exec(unescape(block.replace(TAG, " ")));
    comps.push({ title, price: Number(money[0].replace(/,/g, "")), source: "ebay_sold", url: link ? link[1].split("?")[0] : url, date: d ? d[1] : "", note: "" });
    if (comps.length >= limit) break;
  }
  return comps;
}

let browseToken: { token: string; exp: number } | null = null;

async function getBrowseToken(id: string, secret: string): Promise<string | null> {
  if (browseToken && browseToken.exp > Date.now() + 60_000) return browseToken.token;
  try {
    const res = await fetch("https://api.ebay.com/identity/v1/oauth2/token", {
      method: "POST",
      headers: { "content-type": "application/x-www-form-urlencoded", authorization: "Basic " + Buffer.from(`${id}:${secret}`).toString("base64") },
      body: "grant_type=client_credentials&scope=" + encodeURIComponent("https://api.ebay.com/oauth/api_scope"),
      signal: AbortSignal.timeout(10000),
    });
    if (!res.ok) return null;
    const j = (await res.json()) as { access_token: string; expires_in: number };
    browseToken = { token: j.access_token, exp: Date.now() + j.expires_in * 1000 };
    return j.access_token;
  } catch {
    return null;
  }
}

/** Active listings via the official Browse API — asking prices, flagged as such. */
export async function fetchActiveComps(query: string, clientId: string, clientSecret: string, limit = 20): Promise<Comp[]> {
  if (!query.trim()) return [];
  const token = await getBrowseToken(clientId, clientSecret);
  if (!token) return [];
  try {
    const url = `https://api.ebay.com/buy/browse/v1/item_summary/search?q=${encodeURIComponent(query)}&limit=${limit}&filter=${encodeURIComponent("buyingOptions:{FIXED_PRICE|AUCTION},conditions:{USED|NEW}")}`;
    const res = await fetch(url, { headers: { authorization: `Bearer ${token}`, "X-EBAY-C-MARKETPLACE-ID": "EBAY_US" }, signal: AbortSignal.timeout(10000), cache: "no-store" });
    if (!res.ok) return [];
    const j = (await res.json()) as { itemSummaries?: { title: string; price?: { value: string }; itemWebUrl: string; condition?: string }[] };
    return (j.itemSummaries ?? [])
      .filter((i) => i.price?.value)
      .map((i) => ({ title: i.title, price: Number(i.price!.value), source: "ebay_active", url: i.itemWebUrl, date: "", note: `asking price${i.condition ? ", " + i.condition : ""}` }));
  } catch {
    return [];
  }
}

export function summarize(comps: Comp[]): { n: number; p25?: number; median?: number; p75?: number } {
  const prices = comps.map((c) => c.price).filter((p) => p > 0).sort((a, b) => a - b);
  if (!prices.length) return { n: 0 };
  const q = (f: number) => prices[Math.min(prices.length - 1, Math.floor(f * (prices.length - 1)))];
  return { n: prices.length, p25: q(0.25), median: q(0.5), p75: q(0.75) };
}
